pythran/analyses/range_values.py

```diff
@@ -0,0 +1,448 @@
+        """ Update range value for augassigned variables.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse("def foo(): a = 2; a -= 1")
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=1, high=2)
+        """
+        if isinstance(node.target, ast.Name):
+            low, high = self.combine(self.result[node.target.id],
+                                     self.visit(node.value),
+                                     node.op)
+            self.add(node.target.id, low, high)
+
+    def visit_For(self, node):
+        """ Handle iterate variable in for loops.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse('''
+        ... def foo():
+        ...     a = b = c = 2
+        ...     for i in range(1):
+        ...         a -= 1
+        ...         b += 1''')
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=-inf, high=2)
+        >>> res['b']
+        Range(low=2, high=inf)
+        >>> res['c']
+        Range(low=2, high=2)
+        """
+        assert isinstance(node.target, ast.Name), "For apply on variables."
+        self.add(node.target.id, *self.visit(node.iter))
+        old_range = copy.deepcopy(self.result)
+        map(self.visit, node.body)
+        for name, (low, high) in old_range.iteritems():
+            if self.result[name].low < low:
+                self.result[name] = \
+                    self.result[name]._replace(low=-float("inf"))
+            if self.result[name].high > high:
+                self.result[name] = \
+                    self.result[name]._replace(high=float("inf"))
+        map(self.visit, node.orelse)
+
+    def visit_While(self, node):
+        """ Handle incremented variables in loop body.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse('''
+        ... def foo():
+        ...     a = b = c = 2
+        ...     while a > 0:
+        ...         a -= 1
+        ...         b += 1''')
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=-inf, high=2)
+        >>> res['b']
+        Range(low=2, high=inf)
+        >>> res['c']
+        Range(low=2, high=2)
+        """
+        old_range = copy.deepcopy(self.result)
+        map(self.visit, node.body)
+        for name, (low, high) in old_range.iteritems():
+            if self.result[name].low < low:
+                self.result[name] = \
+                    self.result[name]._replace(low=-float("inf"))
+            if self.result[name].high > high:
+                self.result[name] = \
+                    self.result[name]._replace(high=float("inf"))
+        map(self.visit, node.orelse)
```

Comment:
I am not very happy with the formalism. I think your equation looks like

Ranges(after_while) = Range(before_while) U Range(while_body) *

where the * operator is a widening operator. The equation is not as accurate as it could be (you're missing the condition) and it's not explicit. You could hade a widen operator to the Range class ;-)